src/idl/tests/forward.c

```diff
@@ -0,0 +1,153 @@
+    { IDL_RETCODE_SEMANTIC_ERROR, 0u, "union a; struct a; struct a { long b; };" },
+    { IDL_RETCODE_SEMANTIC_ERROR, 0u, "union a; struct a; union a switch (short) { case 1: long b; };" },
+    { IDL_RETCODE_SYNTAX_ERROR, 0u, "enum a; enum a { b, c };" }
+  };
+
+  for (size_t i = 0; i < sizeof (tests) / sizeof (tests[0]); i++) {
+    idl_retcode_t ret;
+    idl_pstate_t *pstate = NULL;
+    printf("test idl: %s\n", tests[i].idl);
+    ret = parse_string(tests[i].idl, &pstate);
+    CU_ASSERT_EQUAL_FATAL(ret, tests[i].retcode);
+    if (ret == IDL_RETCODE_OK) {
+      const idl_forward_t *forward = NULL;
+      const idl_type_spec_t *node;
+      CU_ASSERT_PTR_NOT_NULL_FATAL(pstate);
+      CU_ASSERT_EQUAL_FATAL(ret, IDL_RETCODE_OK);
+      assert(pstate);
+      for (node = pstate->root; idl_is_forward(node); node = idl_next(node))
+        forward = node;
+      CU_ASSERT_FATAL(idl_is_forward(forward));
+      assert(forward);
+      CU_ASSERT_EQUAL(idl_type(node), tests[i].type);
+      CU_ASSERT_PTR_EQUAL(forward->type_spec, node);
+    }
+    idl_delete_pstate(pstate);
+  }
+}
+
+CU_Test(idl_forward, aliases)
+{
+  static const struct {
+    size_t forwards;
+    const char *idl;
+    idl_retcode_t retcode;
+    idl_type_t type;
+  } tests[] = {
+    { 1, "struct a; typedef a b; struct a { long b; };", IDL_RETCODE_OK, IDL_STRUCT },
+    { 2, "struct a; struct a; typedef a b; struct a { long b; };", IDL_RETCODE_OK, IDL_STRUCT },
+    { 1, "union a; typedef a b; union a switch(short) { case 1: long b; };", IDL_RETCODE_OK, IDL_UNION },
+    { 2, "union a; union a; typedef a b; union a switch(short) { case 1: long b; };", IDL_RETCODE_OK, IDL_UNION }
+  };
+
+  for (size_t i = 0; i < sizeof(tests) / sizeof(tests[0]); i++) {
+    idl_retcode_t ret;
+    idl_pstate_t *pstate = NULL;
+    printf("test idl: %s\n", tests[i].idl);
+    ret = parse_string(tests[i].idl, &pstate);
+    CU_ASSERT_EQUAL(ret, tests[i].retcode);
+    if (ret == IDL_RETCODE_OK) {
+      idl_forward_t *forward[2];
+      idl_typedef_t *alias;
+      idl_type_spec_t *type_spec;
+      size_t n = tests[i].forwards - 1;
+      for (size_t j = 0; j < tests[i].forwards; j++) {
+        forward[j] = j ? idl_next(forward[j - 1]) : (idl_forward_t *)pstate->root;
+        CU_ASSERT_FATAL(idl_is_forward(forward[j]));
+      }
+      alias = idl_next(forward[n]);
+      CU_ASSERT_FATAL(idl_is_typedef(alias));
+      type_spec = idl_next(alias);
+      CU_ASSERT_EQUAL_FATAL(idl_type(type_spec), tests[i].type);
+      CU_ASSERT_PTR_EQUAL(forward[n]->type_spec, type_spec);
+      CU_ASSERT_PTR_EQUAL(alias->type_spec, forward[n]);
+    }
+    idl_delete_pstate(pstate);
+  }
+}
+
+CU_Test(idl_forward, backwards_aliases)
+{
+  static const struct {
+    size_t forwards;
+    const char *idl;
+    idl_retcode_t retcode;
+    idl_type_t type;
+  } tests[] = {
+    { 1, "struct a { long b; }; struct a; typedef a b;", IDL_RETCODE_OK, IDL_STRUCT },
```

Comment:
Well, in the spec an example is provided just below:

```
module M {
  typedef short M;
  interface I {
    void i (in short j);
  };
};
```

I based the implementation of scoping rules on that, my reasoning being naming conflicts with constructors in e.g. C++. In this particular example `i` wouldn't clash with `I` in C++ (case sensitive), but since IDL is case-insensitive, comparison of identifiers must happen in a case insensitive manner and therefore they clash(?)

Reply:
@k0ekk0ek I still take the position that the spec is unclear because it doesn't state anywhere what is included in a definition, except in the grammar and there a struct member is not a definition. Moreover, there is a fundamental difference between types and data and so it is also perfectly reasonable that the two would be treated differently.

That said, you are correct that C++ screwed things up, I forgot about its tendency to do so. Whether the language should be faulted or the OMG for devising a new IDL-to-C++ mapping that made previously legal code illegal, I don't know. I'll be fair and blame them both 🙂 because in the old C++ mapping (from 1999) there was no problem:
```
struct a { long a; };
```
was mapped to essentially identical C++ code. No constructors, just a public member `a` of a `struct` named `a`, and that works just fine. Then comes along the C++11 mapping, which indeed produces a `int32_t a()` accessor that then causes:
```
xxx.h:56:10: error: constructor cannot have a return type
    int32_t m() const { return this->m_; }
    ~~~~ ^
```
So it really doesn't matter anymore whether it was meant to work ...

As to the equivalence of 
```
struct a { long b; }; typedef a b;
struct b { long b; }; typedef b a;
```
I'd say the IDL spec requires them to be treated the same, and so they should both be accepted or both be rejected. In C++ they end up being very different because a typedef doesn't introduce new constructors — so much for abstraction ... — and since C++ seems to be the limiting case, it is not worth holding up this PR over this detail.